### PR TITLE
ARTEMIS-3057 Add min-disk-free feature

### PR DESCRIPTION
**Pull request by haanhvu**
fixes https://issues.apache.org/jira/browse/ARTEMIS-3057

@clebertsuconic please review

### CODE REVIEW

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/GlobalDiskFullTest.java [hunk 1]
```diff
@@ -51,11 +51,19 @@ public void tick(long usableSpace, long totalSpace) {
          }
 
          @Override
-         public void over(long usableSpace, long totalSpace) {
```
[the remark this answers is not in the document]
Comment (haanhvu) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/GlobalDiskFullTest.java:54]: @clebertsuconic for max-disk-usage, over means violate. But for min-disk-free, over means ok (not violate).

I know changing a method name of a public interface can be a flag. But I did it for the sake of clarification.

**Comment by haanhvu** [comment]
I'm amending the PR as we discussed. Please wait.

**Comment by haanhvu** [comment]
@clebertsuconic @gemmellr @jbertram 

I amended the PR as we discussed. Please review. (Sorry I had an important certificate exam last week, so I didn't have much time on the PR).

I got `BUILD SUCCESS` with `mvn -Pdev install -Derrorprone`. Not sure if it's the same with the Github workflow. Please help activate the workflow.

Also, I feel like I need to add more tests, especially the configuration tests, for the overriding behaviors. But I'm not sure which test cases should be prioritized here. Any advice?

**Comment by clebertsuconic** [comment]
@haanhvu Perhaps Wildfly is using this API on their integration. I'm adding a commit to deprecate the older method and adding a test. 

I'm just going to run the complete test suite on my CI. and I will look into merge this soon.

**Comment by clebertsuconic** [comment]
Can you check on testSimpleTickforMinDiskFree(). it's always failing.

**Comment by clebertsuconic** [comment]
another real failure is 
org.apache.activemq.artemis.tests.integration.jms.cluster.TopicClusterPageStoreSizeTest.testPageStoreSizeWithClusteredDurableSub


I will check this later...

(If you could check that one please?)


But again, for the purpose of your Outreachy application you should consider your task done... 

I'm just asking you now in the spirit of open source :) ... I will also look for why these failures tomorrow.

**Comment by haanhvu** [comment]
> Can you check on testSimpleTickforMinDiskFree(). it's always failing.

That test is just a following copy of testSimpleTickforMaxDiskUsage() (initally testSimpleTick())

The test fails because I haven't fully understood the logic of testSimpleTick. Can you pls explain to me the logic of this code:
```
      storeMonitor.tick();

      Assert.assertEquals(0, overMaxUsage.get());
      Assert.assertEquals(1, tick.get());
      Assert.assertEquals(1, underMaxUsage.get());

      storeMonitor.setMaxUsage(0);

      storeMonitor.tick();

      Assert.assertEquals(1, overMaxUsage.get());
      Assert.assertEquals(2, tick.get());
      Assert.assertEquals(1, underMaxUsage.get());
```

**Comment by haanhvu** [comment]
> another real failure is org.apache.activemq.artemis.tests.integration.jms.cluster.TopicClusterPageStoreSizeTest.testPageStoreSizeWithClusteredDurableSub
> 
> I will check this later...
> 
> (If you could check that one please?)
> 
> But again, for the purpose of your Outreachy application you should consider your task done...
> 
> I'm just asking you now in the spirit of open source :) ... I will also look for why these failures tomorrow.

this also, what does this test do?

**Comment by gemmellr** [comment]
> > Can you check on testSimpleTickforMinDiskFree(). it's always failing.
> 
> That test is just a following copy of testSimpleTickforMaxDiskUsage() (initally testSimpleTick())
> 
> The test fails because I haven't fully understood the logic of testSimpleTick. Can you pls explain to me the logic of this code:
> 
> ```
>       storeMonitor.tick();
> 
>       Assert.assertEquals(0, overMaxUsage.get());
>       Assert.assertEquals(1, tick.get());
>       Assert.assertEquals(1, underMaxUsage.get());
> 
>       storeMonitor.setMaxUsage(0);
> 
>       storeMonitor.tick();
> 
>       Assert.assertEquals(1, overMaxUsage.get());
>       Assert.assertEquals(2, tick.get());
>       Assert.assertEquals(1, underMaxUsage.get());
> ```

The existing test creates a storeMonitor and passes a maxUsage value of 0.999, such that its checking a limit of 99.9% usage, i.e something it assumes it should hopefully be under. After the first tick shown there, it checks the 'over' callback was not called and that the 'under' callback was. It then alters the maxUsage value to be 0, meaning 0% usage, which it assumes the file it wrote originally plus any other existing usage, will hopefully push it usage over the new low limit. It then calls tick again and ensures the 'over' callback fired but not the 'under' callback. (Its assumptions are a little brittle, but will typically work).

Your new test would do something similar but use extreme values to compare against the actual free space limit, say 0/1 byte and Long.MAX_VALUE, to try and ensure the actual value went over/under it.

**Comment by gemmellr** [comment]
> I got `BUILD SUCCESS` with `mvn -Pdev install -Derrorprone`. Not sure if it's the same with the Github workflow. Please help activate the workflow.
> 

I cant approve the workflow here, but you can run it in your own fork regardless by enabling the actions there, at e.g https://github.com/haanhvu/activemq-artemis/actions. Then if you push to a branch which is not already in use for a PR then only your fork will run the workflow, and you can check the results etc there. Then you can raise a PR when ready (or update an existing branch already used for an open PR, e.g this one).

**Comment by haanhvu** [comment]
> The existing test creates a storeMonitor and passes a maxUsage value of 0.999, such that its checking a limit of 99.9% usage, i.e something it assumes it should hopefully be under. After the first tick shown there, it checks the 'over' callback was not called and that the 'under' callback was. It then alters the maxUsage value to be 0, meaning 0% usage, which it assumes the file it wrote originally plus any other existing usage, will hopefully push it usage over the new low limit. It then calls tick again and ensures the 'over' callback fired but not the 'under' callback. (Its assumptions are a little brittle, but will typically work).

@gemmellr Thanks a lot! I tweaked the test and it passed.

@clebertsuconic can you review this again? It passed on my own fork.



> another real failure is
> org.apache.activemq.artemis.tests.integration.jms.cluster.TopicClusterPageStoreSizeTest.testPageStoreSizeWithClusteredDurableSub

I didn't even touch this. Not sure why it passes now :)

Do you think we need to add more tests?

**Comment by chibenwa** [comment]
> I will merge this right after 2.20 is released...

Hello @clebertsuconic .

As far as I can see 2.20 is released. Do you still plan on merging this work?

**Comment by jbertram** [comment]
@clebertsuconic, it's been a year since this was apparently ready to be merged. Let me know if you still think it should be merged and I'll resolve the conflicts and merge it. Thanks!

**Comment by jbertram** [comment]
@clebertsuconic, it's been almost two years since this was apparently ready to be merged. Let me know if you still think it should be merged and I'll resolve the conflicts and merge it. Thanks!

**Comment by clebertsuconic** [comment]
@jbertram I think it needs some testing and reviewing. but if you could handle it? don't assume it's ready to merge without reviewing it.

**Comment by clebertsuconic** [comment]
@jbertram yeah.. but then I was not sure if it was 100% tested. and I never got to do it.

let me know if you're doing this... otherwise I will make sure I do it :)

**Comment by jbertram** [comment]
I'm resolving the conflicts and refactoring a few changes that I'm not sold on. I'll probably close this PR and send a new one, but the author will still be @haanhvu.

**Comment by jbertram** [comment]
Superseded by #4599.

**Pull request closed without merging**